lib/action_view/component/base.rb

```diff
@@ -110,34 +110,40 @@ def call_method_name(variant)
           end
         end
 
-        def has_initializer?
-          self.instance_method(:initialize).owner == self
-        end
-
         def source_location
-          # Require #initialize to be defined so that we can use
-          # method#source_location to look up the file name
-          # of the component.
-          #
-          # If we were able to only support Ruby 2.7+,
-          # We could just use Module#const_source_location,
-          # rendering this unnecessary.
-          raise NotImplementedError.new("#{self} must implement #initialize.") unless has_initializer?
-
-          instance_method(:initialize).source_location[0]
+          @source_location ||=
+              begin
+                # Require #initialize to be defined so that we can use
+                # method#source_location to look up the file name
+                # of the component.
+                #
+                # If we were able to only support Ruby 2.7+,
+                # We could just use Module#const_source_location,
+                # rendering this unnecessary.
+                #
+                initialize_method = instance_method(:initialize)
+                initialize_method.source_location[0] if initialize_method.owner == self
+              end
         end
 
         def compiled?
           @compiled && ActionView::Base.cache_template_loading
         end
 
+        def compile!
+          compile(validate: true)
+        end
+
         # Compile templates to instance methods, assuming they haven't been compiled already.
         # We could in theory do this on app boot, at least in production environments.
         # Right now this just compiles the first time the component is rendered.
-        def compile
+        def compile(validate: false)
           return if compiled?
 
-          validate_templates
+          if template_errors.present?
+            raise ActionView::Component::TemplateError.new(template_errors) if validate
+            return false
+          end
```

Comment:
The logic of compilation is much easier to grok now. Plus we can report on all the mistakes you made with the component in one go.